relayer/chains/substrate/provider.go

```diff
@@ -0,0 +1,468 @@
+package substrate
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"path"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	transfertypes "github.com/cosmos/ibc-go/v5/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v5/modules/core/02-client/types"
+	conntypes "github.com/cosmos/ibc-go/v5/modules/core/03-connection/types"
+	chantypes "github.com/cosmos/ibc-go/v5/modules/core/04-channel/types"
+	ibcexported "github.com/cosmos/ibc-go/v5/modules/core/exported"
+	"github.com/cosmos/relayer/v2/relayer/chains/substrate/keystore"
+	"github.com/cosmos/relayer/v2/relayer/provider"
+	"github.com/gogo/protobuf/proto"
+	"go.uber.org/zap"
+)
+
+var (
+	_ provider.ChainProvider  = &SubstrateProvider{}
+	_ provider.KeyProvider    = &SubstrateProvider{}
+	_ provider.ProviderConfig = &SubstrateProviderConfig{}
+)
+
+type SubstrateProviderConfig struct {
+	Key            string `json:"key" yaml:"key"`
+	ChainName      string `json:"-" yaml:"-"`
+	ChainID        string `json:"chain-id" yaml:"chain-id"`
+	RPCAddr        string `json:"rpc-addr" yaml:"rpc-addr"`
+	RelayRPCAddr   string `json:"relay-rpc-addr" yaml:"rpc-addr"`
+	AccountPrefix  string `json:"account-prefix" yaml:"account-prefix"`
+	KeyringBackend string `json:"keyring-backend" yaml:"keyring-backend"`
+	KeyDirectory   string `json:"key-directory" yaml:"key-directory"`
+	Debug          bool   `json:"debug" yaml:"debug"`
+	Timeout        string `json:"timeout" yaml:"timeout"`
+	Network        uint8  `json:"network" yaml:"network"`
+}
+
+func (spc SubstrateProviderConfig) Validate() error {
+	if _, err := time.ParseDuration(spc.Timeout); err != nil {
+		return fmt.Errorf("invalid Timeout: %w", err)
+	}
+	return nil
+}
+
+func keysDir(home, chainID string) string {
+	return path.Join(home, "keys", chainID)
+}
+
+// NewProvider validates the SubstrateProviderConfig, instantiates a ChainClient and then instantiates a SubstrateProvider
+func (spc SubstrateProviderConfig) NewProvider(log *zap.Logger, homepath string, debug bool, chainName string) (provider.ChainProvider, error) {
+	spc.KeyDirectory = keysDir(homepath, spc.ChainID)
+	sp := &SubstrateProvider{
+		PCfg: spc,
+	}
+
+	err := sp.Init()
+	if err != nil {
+		return nil, err
+	}
+
+	return sp, nil
+}
+
+func (sp *SubstrateProvider) Init() error {
+	keybase, err := keystore.New(sp.PCfg.ChainID, sp.PCfg.KeyringBackend, sp.PCfg.KeyDirectory, sp.Input)
+	if err != nil {
+		return err
+	}
+
+	sp.Keybase = keybase
+	return nil
+}
+
+type SubstrateProvider struct {
+	log     *zap.Logger
```

Comment:
I think a new log is initialized for the cosmos provider because it uses the ChainClient implementation from lens. We should be able to just pass the log to the SubstrateProvider like here right? https://github.com/cosmos/relayer/blob/d59c98adb368822a25328fe74db433f8aae78a1d/relayer/chains/cosmos/provider.go#L65